File: python-webdriver-tests/sort_condition_provider.py
from sort_criteria import SortCriteria

class SortConditionProvider:

    def __init__(self, columns):
        self.columns = list(columns)

    def product_names(self):
        items = []
        for i in range(len(self.columns)):
            items = items + self.product(items, self.columns, True)

        return self.uniq(items)

    def add_directs(self, names):

        sort_directs = ['asc', 'desc']
        all_directs = []
        sortConditions = []
        for i in range(len(names)):
            all_directs = self.product(all_directs, sort_directs)
        for directs in all_directs:
            sortCondition = [{"sortName": names[idx], "sortDirect": direct} for idx, direct in enumerate(directs)]
            sortConditions.append(sortCondition)
        return sortConditions

    def product(self, first_arr, second_arr, uniq=False):
        if(not first_arr):
            return [[i] for i in second_arr]
        result = []
        for first in first_arr:
            for second in second_arr:
                item = first[:]
                if(not uniq or not second in item):
                    item.append(second)
                    result.append(item)
        return result

    def uniq(self, arr):
        result = []
        for item in arr:
            if(not item in result):
                result.append(item)
        return result

    def all(self):
        conditions = []
        names_arr = self.product_names()
        for names in names_arr:
            conditions += self.add_directs(names)
        return [SortCriteria(condition) for condition in conditions];

File: lib/ember-cli-stub-server/lib/stubs/sort-condition-provider.js
var helper = require('./helper');
var SortCondition = require('./sort-condition')
var sortDirects = ['asc', 'desc'];

var SortConditionProvider = function (columns){
  this.sortColumns =  columns || [];
};

SortConditionProvider.prototype.forEach = function (callBack) {
  var namesArr = this.productNames();
  var self = this;
  namesArr.forEach(function (names) {
    self.mapDirects(names).forEach(function (directs) {
      callBack(new SortCondition(names, directs));
    });
  });
};

SortConditionProvider.prototype.productNames = function () {
  var result = [];
  var self = this;
  this.sortColumns.forEach(function() {
    result = result.concat(self.product(result, self.sortColumns, true));
  });
  return helper.uniq(result);
};

SortConditionProvider.prototype.mapDirects = function (names) {
  var sortDirects = ['asc', 'desc'];
  var allDirects = [];
  var self = this;
  names.forEach(function () {
    allDirects = self.product(allDirects, sortDirects);
  });
  return allDirects;
};


SortConditionProvider.prototype.product = function (firstArr, secondArr, uniq) {
  if(firstArr.length === 0){
    return secondArr.map(function (item) {
      return [item];
    });
  }
  var result = [];
  firstArr.forEach(function (first) {
    secondArr.forEach(function (second) {
      var item = first.slice();
      if(!uniq || item.indexOf(second) === -1){
        item.push(second);
        result.push(item);
      }
    });
  });
  return result;
};





module.exports = SortConditionProvider;
